### PR TITLE
azd package support for user specified output paths

### LINKED ISSUE #2543: [Request] azd package writes output folder to an env variable
Azd package is quickly becoming one of my favorite tools to easily and reliably create the app package build before provisioning and deployment.  It's useful for automation and diagnostics.

That said, to better enable pipelines and tools, I request the output path for *this package gets written to an environment variable or something that can be looked up in scripts/code/pipelines.  Another alternative I suppose is a flag that enables reliable output to a known folder.

@jongio @ellismg @savannahostrowski

### PR DESCRIPTION
**Pull request by wbreza**
Resolves #2543

Adds ability for a user to specify a `--output-path` param to the `azd package` command to control the output location of file based packages.

For top level commands it would be expected that user specify a absolute or relative folder path.  If the path does not exist azd will create it automatically.

### Output multiple services to well known location
`azd package --output-path ./dist`

image

### Output single service to well known location
`azd package api --output-path ./dist/api-app.zip`

image

### Updates default naming convention for package artifacts

Package artifact naming convention now follows the format of `{project_name}-{service_name}-azddeploy-{timestamp}.zip` to provide more information to the user to more easily find the correct generated artifact and distinguish them from one another.

image

## Use Case: Store package output into some artifact store

When running AZD in CI a common scenario would be to package service artifacts into a well known location that can then be referenced and stored in an artifact store.

User can run the following:
`azd package api --output-path ./dist/api-app.zip`

Since the user now has control of where the artifact will be generated they can easily copy this artifact into an artifact store of their choosing.

## Use Case: Deploy azd package from previously packaged artifact

User now has more control on where the generated artifacts are stored and can easily reference them in other azd commnds such as `azd deploy`

```bash
// Generate service artifact to known path
azd package api --output-path ./dist/api-app.zip

// Use known path in deploy command
azd deploy api --from-package ./dist/api-app.zip
```

TODO: 
- [x] Write functional test to validate behavior

### CODE REVIEW

**Review by ellismg — APPROVED** [review]
LGTM - I wonder if for the case where you want to publish all your deployment packages, we should revisit our naming strategy for the packages themselves.  Perhaps we could bake in the service name (and maybe remove the unique suffix) so you ended up with package names like `azd-service-web.zip` or something that makes it a little more clear which project things are tied back to?

As part of this - are we comfortable with the artifact we produce for containerized applications?  We sort of punted on what a package for a containerized application is, and IIRC instead used something more like "here's the name of the local or remote docker image that you can run".  Should we be exporting this image in some way in this case?  Generating a manifest that `azd` can understand that contains just the fully qualified image id?  Something else?

**Comment by wbreza** [comment]
@Austinauth / @savannahostrowski Any feedback on UX here? Nothing net new just illustrating that the output path would be that defined by the user vs our default strategy if they opt-in to the `--output-path` param.

**Comment by savannahostrowski** [comment]
So does this override the conventions that might be outlined in a package.json or other similar file in another language? Just wondering how explicitly setting this plays with other files that define the output dir in the user's project already. What takes precedence?

**Comment by wbreza** [comment]
> So does this override the conventions that might be outlined in a package.json or other similar file in another language? Just wondering how explicitly setting this plays with other files that define the output dir in the user's project already. What takes precedence?

@savannahostrowski The package artifacts sit on top of the any services `dist` setting.  The `dist` setting today is used for intermediate build artifacts which are then zipped up into the uber package artifact.

Today all `azd package` commands write to the temp folder, this PR only changes this behavior to copy the generated zip artifact to a user specified location via command param `--output-path`.

**Review by vhvb1989 — APPROVED** [review]
Added a comment about the options to include *ServiceBuildResult

**Review by savannahostrowski — APPROVED** [review]

File: cli/azd/pkg/project/service_manager.go [hunk 1]
```diff
@@ -361,6 +372,45 @@ func (sm *serviceManager) Package(
...
+		// Package path can be a file path or a container image name
+		// We only move to desired output path for file based packages
+		_, err = os.Stat(packageResult.PackagePath)
+		hasPackageFile := err == nil
+
+		if hasPackageFile && options.OutputPath != "" {
+			var destFilePath string
+			var destDirectory string
+
+			isFilePath := filepath.Ext(options.OutputPath) != ""
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/project/service_manager.go:384]: Today all the generated artifacts are zip files so they would always have an extension.  If we added additional packaging options then we could think about this.

File: cli/azd/pkg/project/service_manager.go [hunk 2]
```diff
@@ -361,6 +372,45 @@ func (sm *serviceManager) Package(
 			return
 		}
 
+		// Package path can be a file path or a container image name
+		// We only move to desired output path for file based packages
+		_, err = os.Stat(packageResult.PackagePath)
+		hasPackageFile := err == nil
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/project/service_manager.go:378]: This feature would only target service targets that support zip deploy.  It is unlikely a generated image path would exist as a file in the local system.

File: cli/azd/pkg/project/service_manager.go [hunk 3]
```diff
@@ -68,6 +73,7 @@ type ServiceManager interface {
 		ctx context.Context,
 		serviceConfig *ServiceConfig,
 		buildOutput *ServiceBuildResult,
+		options *PackageOptions,
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/project/service_manager.go:76]: Thanks for the feedback.  I will think about this.

**Pull request merged**